### ISSUE TITLE
MSL: Incorrect IO Structure and Member Names for Vertex Shaders Compiled using CompilerMSL::Options::vertex_for_tessellation=true

### ISSUE THREAD

**Issue by NicVanZuylen** [KhronosGroup/SPIRV-Cross#1496]
I'm attempting to compile a vertex shader from HLSL->SPIRV->MSL and compiling the output into an MTLLibrary will throw the following errors:

```
program_source:27:28: error: redefinition of 'out' with a different type: 'device VsPosColor_out &' vs 'VsPosColor_out'
    device VsPosColor_out& out = spvOut[gl_GlobalInvocationID.y * spvStageInputSize.x + gl_GlobalInvocationID.x];
                           ^
program_source:26:20: note: previous definition is here
    VsPosColor_out out = {};
                   ^
program_source:30:9: error: no member named 'gl_Position' in 'VsPosColor_out'
    out.gl_Position = float4(in.in.in_var_POSITION, 0.0, 1.0);
    ~~~ ^
program_source:30:33: error: no member named 'in' in 'VsPosColor_in_1'
    out.gl_Position = float4(in.in.in_var_POSITION, 0.0, 1.0);
                             ~~ ^
program_source:31:9: error: no member named 'out' in 'VsPosColor_out'
    out.out.out_var_COLOR0 = in.in.in_var_COLOR0;
    ~~~ ^
program_source:31:33: error: no member named 'in' in 'VsPosColor_in_1'
    out.out.out_var_COLOR0 = in.in.in_var_COLOR0;
                             ~~ ^
program_source:
```

With MSL source output from SPIRV-Cross:
```
#include <metal_stdlib>
#include <simd/simd.h>

using namespace metal;

struct VsPosColor_in
{
    float2 in_var_POSITION [[attribute(0)]];
    float4 in_var_COLOR0 [[attribute(1)]];
};

struct VsPosColor_out
{
    float4 out_out_var_COLOR0;
    float4 out_gl_Position;
};

struct VsPosColor_in_1
{
    float2 in_in_var_POSITION [[attribute(0)]];
    float4 in_in_var_COLOR0 [[attribute(1)]];
};

kernel void VsPosColor(VsPosColor_in_1 in [[stage_in]], uint3 gl_GlobalInvocationID [[thread_position_in_grid]], uint3 spvStageInputSize [[grid_size]], device VsPosColor_out* spvOut [[buffer(28)]])
{
    VsPosColor_out out = {};
    device VsPosColor_out& out = spvOut[gl_GlobalInvocationID.y * spvStageInputSize.x + gl_GlobalInvocationID.x];
    if (any(gl_GlobalInvocationID >= spvStageInputSize))
        return;
    out.gl_Position = float4(in.in.in_var_POSITION, 0.0, 1.0);
    out.out.out_var_COLOR0 = in.in.in_var_COLOR0;
    return out;
}
```

As you can see from the MSL source, 'out' is defined twice in the kernel function body, and the we are attempting to access member: 'gl_Position' which is undefined. Also, when we access our 'in' and 'out' structures, we seem to access a member of the same name as the structures again before accessing the member we intend to. There is also two versions of 'VsPosColor_in', which seems unnecessary (were these structures intended to be nested?).

**Comment by NicVanZuylen** [comment]
Here is the HLSL source, which may be helpful:
```
struct PSInputPosColor
{
	float4 position : SV_POSITION;
	float4 color : COLOR0;
};

[RootSignature(DefaultRenderSignature)]
PSInputPosColor VsPosColor(float2 position : POSITION, float4 rgba : COLOR0)
{
	PSInputPosColor result;

	result.position = float4(position.xy,0,1);
	result.color = rgba;

	return result;
}
```

**Comment by HansKristian-Work** [comment]
I cannot reproduce this result, I get this with DXC:
```
#include <metal_stdlib>
#include <simd/simd.h>

using namespace metal;

struct VsPosColor_out
{
    float4 out_var_COLOR0;
    float4 gl_Position;
};

struct VsPosColor_in
{
    float2 in_var_POSITION [[attribute(0)]];
    float4 in_var_COLOR0 [[attribute(1)]];
};

kernel void VsPosColor(VsPosColor_in in [[stage_in]], uint3 gl_GlobalInvocationID [[thread_position_in_grid]], uint3 spvStageInputSize [[grid_size]], device VsPosColor_out* spvOut [[buffer(28)]])
{
    device VsPosColor_out& out = spvOut[gl_GlobalInvocationID.y * spvStageInputSize.x + gl_GlobalInvocationID.x];
    if (any(gl_GlobalInvocationID >= spvStageInputSize))
        return;
    out.gl_Position = float4(in.in_var_POSITION, 0.0, 1.0);
    out.out_var_COLOR0 = in.in_var_COLOR0;
}
```

Please upload the exact SPIR-V that is failing for you.

**Comment by NicVanZuylen** [comment]
As it turns out, I just managed to get the correct output (same as yours). To get the incorrect output, I was compiling the shader as a vertex shader, then changing the options to enable Options::vertex_for_tessellation (was disabled for the first compilation) and compiling again.

Most of the options are thier default values for both compilations, but here are the ones I'm explicitly setting:
`Options::set_msl_version(2,1,0)`
`Options::texture_1D_as_2D=true`
`Options::texture_buffer_native=true`
`Options::vertex_for_tessllation=true` (For second compilation)

It seems some metadata from the first compilation might be lingering into the second compilation and corrupting the output. I found CompileMSL::reset() which may help, but its a protected function. I'm not using a released version, I am using code after a commit made on 15/10/2020.
